### PR TITLE
Maintain nondet and dynamic counters across symex instances [blocks: #3976]

### PR DESCRIPTION
**Pull request by tautschnig**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 93ac08fd38367341d8082a744ca8efd1202233d9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98729940

**Review by peterschrammel — APPROVED** [review]
Not for this PR, but maybe worth a performance test in future: Could `util/fresh_symbol` do the job?

File: src/goto-symex/goto_symex.h [hunk 1]
```diff
@@ -418,7 +408,7 @@ class goto_symext
   virtual void symex_input(statet &, const codet &);
   virtual void symex_output(statet &, const codet &);
 
-  symex_nondet_generatort build_symex_nondet;
+  static unsigned nondet_count;
```
Comment (romainbrenguier) [inline src/goto-symex/goto_symex.h]: :-1:

**Review by romainbrenguier — CHANGES_REQUESTED** [review]
I'm not sure why there would be several `goto_symext` instantiations, which need to share symbols. But please don't introduce global variables.

**Comment by tautschnig** [comment]
> I'm not sure why there would be several `goto_symext` instantiations, which need to share symbols. 

cbmc --paths (`path_explorert`) creates a new `goto_symext` instance for each path.

> But please don't introduce global variables.

Ack, but that's just the first commit here, which is a clean revert of your earlier patch. The second commit fixes this. But I can probably also get away without doing that revert and just directly make the modifications done in the second commit.

**Comment by tautschnig** [comment]
@romainbrenguier I have now largely redone this, avoiding any reverts.

@peterschrammel I have followed your suggestion, and that change is now no longer present in here and #3999 takes care of it instead.

**Review by romainbrenguier — APPROVED** [review]

**Comment by romainbrenguier** [comment]
@tautschnig 
> cbmc --paths (path_explorert) creates a new goto_symext instance for each path.
> > But please don't introduce global variables.

> Ack, but that's just the first commit here, which is a clean revert of your earlier patch. The second commit fixes this. But I can probably also get away without doing that revert and just directly make the modifications done in the second commit.

Thanks for the explanation and sorry I didn't get this was fixed in the second commit, I think it's clearer without the revert.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 321402d3247b79f63b9ed7e717c98d8229a45cf9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99235044

**Pull request merged**